go/types/set_iterator_test.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"testing"
+
+	"github.com/attic-labs/testify/assert"
+)
+
+func TestSetIterator(t *testing.T) {
+	assert := assert.New(t)
+
+	numbers := append(generateNumbersAsValues(6), Number(10), Number(15))
+	s := NewSet(numbers...)
+	i := s.Iterator()
+
+	vs := ValueSlice{}
+	for v := i.Next(); v != nil; v = i.Next() {
+		vs = append(vs, v)
+	}
+	assert.True(vs.Equals(numbers), "Expected: %v != actual: %v", numbers, vs)
+
+	i = s.Iterator()
+	assert.Panics(func() { i.AdvanceTo(nil) })
+	assert.Equal(Number(0), i.AdvanceTo(Number(-20)))
```

Comment:
It would be good to test what happens when calling Next after AdvanceTo. E.g. presumably `i.AdvanceTo(5); iterToSlice(i) => [6, 10, 15]`?